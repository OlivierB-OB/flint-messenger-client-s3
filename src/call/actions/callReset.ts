import { ICallResetAction, CALL_RESET } from '../types';

export function callReset(): ICallResetAction {
  return { type: CALL_RESET };
}
