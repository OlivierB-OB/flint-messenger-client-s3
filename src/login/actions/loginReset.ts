import { ILoginResetAction, LOGIN_RESET } from '../types';

export function loginReset(): ILoginResetAction {
  return { type: LOGIN_RESET };
}
