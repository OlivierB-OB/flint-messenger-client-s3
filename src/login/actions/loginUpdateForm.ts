import { ILoginForm, ILoginAction, LOGIN_UPDATE_FORM } from '../types';

export function loginUpdateForm<T extends keyof ILoginForm>(field: T, value: string): ILoginAction {
  return {
    type: LOGIN_UPDATE_FORM,
    field,
    value,
  };
}
