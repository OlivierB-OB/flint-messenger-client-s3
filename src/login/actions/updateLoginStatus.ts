import { ILoginAction, LOGIN_UPDATE_STATUS, ILoginStatus } from '../types';

export function updateLoginStatus(status: ILoginStatus): ILoginAction {
  return {
    type: LOGIN_UPDATE_STATUS,
    status,
  };
}
