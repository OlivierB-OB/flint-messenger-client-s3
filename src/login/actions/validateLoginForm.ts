import { ILoginAction, LOGIN_VALIDATE_FORM } from '../types';

export function validateLoginForm(): ILoginAction {
  return {
    type: LOGIN_VALIDATE_FORM,
  };
}
