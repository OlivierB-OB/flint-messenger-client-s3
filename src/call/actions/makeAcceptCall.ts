import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { action } from '../../utils/action';
import { IAppState } from '../../appReducer';
import { updateCallRemoteStream } from './updateCallRemoteStream';
import { updateCallLocalInputs } from './updateCallLocalInputs';
import { getLocalInputs } from '../utils/getLocalInputs';
import { updateCallPeerConnection } from './updateCallPeerConnection';
import { makeEmit } from '../../realtime/actions/makeEmit';
import { peerConnexionFactory, bindLocalInputToPeerConnexion, assertValidConversationId } from '../utils';
import { makeStartCallMode } from './makeStartCallMode';

export const makeAcceptCall = action((
  conversationId: string,
  target: string,
) => {
  return async (dispatch: ThunkDispatch<IAppState, void, Action>, getState: () => IAppState) => {
    console.log('======================================== START ACCEPT');
    const appState = getState();
    assertValidConversationId(appState, conversationId);

    // start call mode
    dispatch(makeStartCallMode(conversationId, target));

    // Create peer connection
    const peerConnection = peerConnexionFactory(
      (candidate) => dispatch(makeEmit('call-ice-candidate', { conversationId, target, candidate })),
      (stream) => dispatch(updateCallRemoteStream(stream)),
    );
    dispatch(updateCallPeerConnection(peerConnection));

    // Create local media inputs
    const localInputs = await getLocalInputs();
    dispatch(updateCallLocalInputs(localInputs));
    bindLocalInputToPeerConnexion(peerConnection, localInputs);

    // Create an RTC peer connexion offer
    const offer = await peerConnection.createOffer();
    await peerConnection.setLocalDescription(offer);
    dispatch(makeEmit('call-accepted', { conversationId, target, offer: peerConnection.localDescription }));
    console.log('======================================== END ACCEPT')
  };
});
