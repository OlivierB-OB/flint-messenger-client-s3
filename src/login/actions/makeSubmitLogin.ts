import axios from 'axios';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { action } from '../../utils/action';
import { IAppState } from '../../appReducer';
import { history } from '../../history';
import { updateLoginStatus } from './updateLoginStatus';
import { makeInitializeApplication } from '../../layout/actions/makeInitializeApplication';

export const makeSubmitLogin = action(() => {
  return async (dispatch: ThunkDispatch<IAppState, void, Action>, getState: () => IAppState) => {
    dispatch(updateLoginStatus('unavailable'));

    const { login } = getState();
    const { email, password } = login.form;

    // FIXME validate form before sending

    try {
      const response = await axios.post(
        `${process.env.REACT_APP_BACKEND}/login`,
        {
          username: email.value,
          password: password.value,
        },
        { withCredentials: true },
      );
      dispatch(makeInitializeApplication(response.data));
      history.push(`/profile`);
    } catch (error) {
      dispatch(updateLoginStatus('error'));
    }
  };
});
