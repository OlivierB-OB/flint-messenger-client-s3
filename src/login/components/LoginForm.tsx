import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import React from 'react';
import { connect } from 'react-redux';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { ILoginForm } from '../types';
import { loginUpdateForm } from '../actions/loginUpdateForm';
import { makeSubmitLogin } from '../actions/makeSubmitLogin';

interface ILoginFormDisplayProps {
  email: string;
  password: string;
  update<T extends keyof ILoginForm>(field: T, value: string): void;
  submit(): void;
}

function LoginFormDisplay({ email, password, update, submit }: ILoginFormDisplayProps) {
  return (
    <Container maxWidth="xs">
      <Box style={{ margin: '2rem 0' }}>
        <TextField
          label="Email"
          value={email}
          required={true}
          fullWidth={true}
          onChange={(event) => update('email', event.target.value)}
        />
        <TextField
          type="password"
          label="Password"
          value={password}
          fullWidth={true}
          onChange={(event) => update('password', event.target.value)}
        />
      </Box>
      <Box style={{ margin: '2rem 0' }}>
        <Grid container justify="flex-end">
          <Grid item xs={4}>
            <Button color="primary" variant="contained" fullWidth={true} onClick={() => submit()}>
              Submit
            </Button>
          </Grid>
        </Grid>
      </Box>
    </Container>
  );
}

const mapStateToProps = ({ login }: IAppState) => ({
  ...login.form,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<IAppState, void, Action>) => ({
  update: <T extends keyof ILoginForm>(field: T, value: string) => dispatch(loginUpdateForm(field, value)),
  submit: () => dispatch(makeSubmitLogin()),
});

export const LoginForm = connect(mapStateToProps, mapDispatchToProps)(LoginFormDisplay);
