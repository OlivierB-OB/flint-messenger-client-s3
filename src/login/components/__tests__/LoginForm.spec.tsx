import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import { LoginForm } from '../LoginForm';
import { makeSubmitLogin } from '../../actions/makeSubmitLogin';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('LoginForm', () => {
  let Connected: any = LoginForm;
  let spySubmit: jest.Mock;
  let restoreSubmit: () => void;

  beforeEach(() => {
    [spySubmit, restoreSubmit] = makeSubmitLogin.mock(jest.fn());
  });

  afterEach(() => {
    restoreSubmit();
  });

  it('should provide status, email, password and callbacks to LoginFormDisplay', async () => {
    const store = mockStore({
      login: {
        status: 'login status',
        email: 'login email',
        password: 'login password',
      },
    });
    const component = shallow(<Connected store={store} />).dive();
    expect(component).toMatchSnapshot();

    (component.prop('update') as any)('field to update', 'new value');
    expect(store.getActions()).toEqual([
      {
        field: 'field to update',
        value: 'new value',
        type: 'LOGIN_UPDATE_FORM',
      },
    ]);
    store.clearActions();

    (component.prop('submit') as any)();
    expect(spySubmit).toHaveBeenCalled();
  });
});
