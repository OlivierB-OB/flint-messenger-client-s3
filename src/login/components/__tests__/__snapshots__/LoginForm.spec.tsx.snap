// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should provide status, email, password and callbacks to LoginFormDisplay 1`] = `
<LoginFormDisplay
  status="login status"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  submit={[Function]}
  update={[Function]}
/>
`;
